server/analyzer/add_implicit_prefix_lengths.go

```diff
@@ -0,0 +1,121 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/analyzer"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+
+	"github.com/dolthub/doltgresql/server/types"
+)
+
+// defaultIndexPrefixLength is the index prefix length that this analyzer rule applies automatically to TEXT columns
+// in secondary indexes.
+const defaultIndexPrefixLength = 255
+
+// AddImplicitPrefixLengths searches the |node| tree for any nodes creating an index, and plugs in a default index
+// prefix length for any TEXT columns in those new indexes. This rule is intended to be used for Postgres compatibility,
+// since Postgres does not require specifying prefix lengths for TEXT columns.
+func AddImplicitPrefixLengths(_ *sql.Context, _ *analyzer.Analyzer, node sql.Node, _ *plan.Scope, _ analyzer.RuleSelector, _ *sql.QueryFlags) (sql.Node, transform.TreeIdentity, error) {
+	var targetSchema sql.Schema
+	transform.Inspect(node, func(node sql.Node) bool {
+		if st, ok := node.(sql.SchemaTarget); ok {
+			targetSchema = st.TargetSchema().Copy()
+			return false
+		}
+		return true
+	})
+
+	// Recurse through the node tree to fill in prefix lengths. Note that some statements come in as Block nodes
+	// that contain multiple nodes, so we need to recurse through and handle all of them.
+	return transform.Node(node, func(node sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		switch node := node.(type) {
+		case *plan.AddColumn:
+			// For any AddColumn nodes, we need to update the target schema with the column being added, otherwise
+			// we won't be able to find those columns if they are also being added to a secondary index.
+			var err error
+			targetSchema, err = analyzer.ValidateAddColumn(targetSchema, node)
+			if err != nil {
+				return nil, transform.SameTree, err
+			}
+
+		case *plan.CreateTable:
+			newIndexes := make([]*sql.IndexDef, len(node.Indexes()))
+			for i := range node.Indexes() {
+				copy := *node.Indexes()[i]
+				newIndexes[i] = &copy
+			}
+			indexModified := false
+			for _, index := range newIndexes {
+				targetSchema := node.TargetSchema()
+				colMap := schToColMap(targetSchema)
+
+				for i := range index.Columns {
+					col, ok := colMap[strings.ToLower(index.Columns[i].Name)]
+					if !ok {
+						return nil, false, fmt.Errorf("indexed column %s not found in schema", index.Columns[i].Name)
+					}
+					if _, ok := col.Type.(types.TextType); ok && index.Columns[i].Length == 0 {
```

Comment:
This is specifically checking for the `TEXT` type (also, not a big deal, but I generally tend to alias our package as `pgtypes` to make the distinction clear). What about for the other string types? (`VARCHAR`, `NAME`, etc.)

Reply:
I didn't want to specifically check against `pgtypes.TextType` – I started by calling `Type.Type()` and looking for the `TEXT` query type, but several tests started failing because other types return `TEXT` through `Type()` (e.g. `Bool`, `arrayContainer`). Those types returning `TEXT` doesn't seem correct, but I figured checking against `pgtypes.TextType` would be a good initial step. I'm curious if you know why those other types are returning `TEXT`. One implication of this is that Domain types built on Text types won't work in indexes, but if we could check against the query type of `TEXT`, it seems like they would. 

Good point on the `pgtypes` naming. I like being consistent with that and will update. 

The other string types, like `VARCHAR`, are different because they are always stored in the row. It's only `TEXT` here because it's stored out-of-band in an address reference.

Reply:
Regarding the other types returning `TEXT`: it's mostly a remnant of using `sql.Type`, and GMS using it as a quick way to ascertain which type it was given (we have `BaseID` for that). I don't think we use it at all in Doltgres anymore after we've made the change to use the actual return type's OID (at least, we shouldn't be using Vitess types anymore), so perhaps we could change those. However, they don't map to Postgres types since it's mainly a relic of GMS being built for MySQL.